### PR TITLE
Refactor ChunkManager

### PR DESCRIPTION
**Pull request by bryanlb**
* Splits the existing `ChunkManager` into an abstract base class `ChunkManager` and two implementations - `IndexingChunkManager`, `CachingChunkManager`
* Stubs out functionality for the new `CachingChunkManager`, including a basic lifecycle management
* Fixes an issue with `KaldbMetadataStore` that prevented instantiating two metadata stores at the same path, and re-introduces the indexer registration from #110 that was disabled in #121 (f6974c3fbe570d04be1b6da801aa11b327932a41)

### CODE REVIEW

File: kaldb/src/test/java/com/slack/kaldb/chunk/ChunkCleanerTaskTest.java
Comment (bryanlb) [inline kaldb/src/test/java/com/slack/kaldb/chunk/ChunkCleanerTaskTest.java:36]: This was removed in #110, but the test currently still requires this config to be initialized. 

It initially passed tests with this removed because another test that didn't cleanup its config was running before this one. With this refactor the tests are running in a slightly different order, causing this test to now fail.

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java
Comment (vthacker) [inline kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:82]: looks like everywhere we do a `get(0)` ? I guess that's because we have to conform to how the `CachedMetadataStore` hierarchy is?

So like every cache node gets it's znode and whenever the first entry  under it updates to evict/assigned etc we do something

looks fine i guess

Comment (vthacker) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:82]: will there be N top level znodes for every chunk within a cache node?

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java
Comment (vthacker) [inline kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java:63]: this might qualify for a long chain award in terms of LOC :) Anyways this was always there

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:82]: The current design is such that the `CachedMetadataStoreImpl` caches all nested nodes. As we know there exist only a single nested node (`SLOT` in this case) we can safely just get the first item in the list. 

We have a todo to refactor the current metadata store constructors, such that we can initialize the store with data on the top znode, and not have to put the data on a nested node. That will simplify access to these znodes once that's refactored.

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java:63]: https://github.com/slackhq/kaldb/pull/56/files#diff-cf8f5116a98917009ffaee10b4534d5695a235497100e64eab54d90937543d40R283 😄

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/caching/CachingChunkManager.java
Comment (bryanlb) [inline kaldb/src/main/java/com/slack/kaldb/chunk/manager/caching/CachingChunkManager.java:37]: I ended up refactoring this from the original [IntStream on the prior PR](https://github.com/slackhq/kaldb/pull/117/files#diff-0e217760e60aa380caf8feb80e8ab69c44231b5d36a6ebe627fe7d6822d5ee46R57), as the `ReadOnlyChunkImpl` can throw an exception on instantiation. In this specific case we likely would prefer to fail startup of the application instead of allowing some number of slots less than our target to instantiate, as this would never be retried.

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java
Comment (mansu) [inline kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java:30]: Please add a class doc for this class.

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java
Comment (mansu) [inline kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java:78]: I don't think it is valuable to add all the chunk metadata to the tag. I think just adding the chunkId is fine. We can look up metadata store if needed for additional info and add tags as needed. For now, let's drop the rest of the tags. 

Is the query encoded in the trace any where?

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java
Comment (mansu) [inline kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java:107]: On this timeout would we call span.finish? Other wise the span will be flushed after 30 seconds and the span duration will be off.

Also, prefer the duration variant if available.

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/indexing/IndexingChunkManager.java
Comment (mansu) [inline kaldb/src/main/java/com/slack/kaldb/chunk/manager/indexing/IndexingChunkManager.java:39]: Please add a class doc for this class.

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/caching/CachingChunkManager.java
Comment (mansu) [inline kaldb/src/main/java/com/slack/kaldb/chunk/manager/caching/CachingChunkManager.java:16]: Please add a class doc for this class.

File: kaldb/src/test/java/com/slack/kaldb/chunk/ReadOnlyChunkImplTest.java
Comment (mansu) [inline kaldb/src/test/java/com/slack/kaldb/chunk/ReadOnlyChunkImplTest.java:67]: I think it's best to wait until the next diff to merge this class along with it's tests. For now we can remove it.

File: kaldb/src/test/java/com/slack/kaldb/chunk/ReadWriteChunkImplTest.java
Comment (mansu) [inline kaldb/src/test/java/com/slack/kaldb/chunk/ReadWriteChunkImplTest.java:426]: Is there a reason to comment out this part of the unit test?

File: kaldb/src/test/java/com/slack/kaldb/metadata/core/KaldbMetadataStoreTest.java
Comment (mansu) [inline kaldb/src/test/java/com/slack/kaldb/metadata/core/KaldbMetadataStoreTest.java:1398]: Good test!

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java
Comment (mansu) [inline kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:183]: I think we should implement this when we add the caching store.

**Review by mansu — CHANGES_REQUESTED** [review]

File: kaldb/src/main/java/com/slack/kaldb/chunk/Chunk.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/Chunk.java:42]: This method is still available on the `ReadWriteChunkImpl`, just not on the `Chunk` interface as this doesn't make sense for the `ReadOnlyChunkImpl`.

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadWriteChunkImpl.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ReadWriteChunkImpl.java:161]: This shouldn't be a separate command from close, as closing needs to automatically handle cleanup. Per the closeable spec it is required to release all system resources as part of invoking close.

I've refactored this so it's a bit clearer in the `close` method.

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:178]: Cleanup has to happen at the very end of the `ReadOnlyChunkImpl`, prior to exiting. We need to first shut down the logsearch, executor service, and deregister our node from ZK prior to deleting the data off of disk.

This ensures that we stop any new inbound queries, and stop any pending queries before removing the lucene index files.

File: kaldb/src/main/java/com/slack/kaldb/chunk/Chunk.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/Chunk.java:30]: Done.

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java:1]: I ended up removing one layer of nesting, such that all `ChunkManager` related classes are in a package, and all `Chunk` related classes are in a package.

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:74]: Let's address the metadata store refactor in a separate diff, as this one is already getting quite large.

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java:78]: These were added as part of #91, and were exceptionally useful when troubleshooting the performance issue we had with the Armeria threadpools.

You can see various screenshots on that PR of the tags we added to the traces. The chunkID is helpful when troubleshooting as that directly maps to the blob store directory.

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:183]: This was addressed as part of a fix above - https://github.com/slackhq/kaldb/pull/122#discussion_r715809724

File: kaldb/src/test/java/com/slack/kaldb/chunk/ReadWriteChunkImplTest.java
Comment (bryanlb) [reply kaldb/src/test/java/com/slack/kaldb/chunk/ReadWriteChunkImplTest.java:426]: Ah yes, I've now removed this. This test used the previous stub of `ReadOnlyChunkImpl`, which allowed instantiation of a `ReadOnlyChunkImpl` against a directory on disk. This is no longer possible as the `ReadOnlyChunkImpl` is now wired against ZK events.

All of the tests for verifying the `ReadOnlyChunkImpl` can hydrate / load / query assets from S3 were moved into the `ReadOnlyChunkImplTest` as part of #124 - (https://github.com/slackhq/kaldb/blob/6f578681ae137b90654ee10cdd47745fbc8e1d61/kaldb/src/test/java/com/slack/kaldb/chunk/ReadOnlyChunkImplTest.java)

File: kaldb/src/test/java/com/slack/kaldb/chunk/ReadOnlyChunkImplTest.java
Comment (bryanlb) [reply kaldb/src/test/java/com/slack/kaldb/chunk/ReadOnlyChunkImplTest.java:67]: Removed

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:91]: This has already been addressed as part of #124 - https://github.com/slackhq/kaldb/pull/124/files#diff-08ae1b7f6d010091ac19c82949597708014c27300d7dc2352525ba07e7011c24R150

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:193]: Done.

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:162]: Done.

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:67]: This has been addressed in #124 to use `SearchContext`.

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java:37]: Let's do this as a separate PR, as this is existing code that wasn't changed in this diff. I don't want to expand the scope with other refactors as this is already quite large.

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:81]: Let's discuss this more in #124 as that has a working implementation reacting to state transitions.

I think reacting on the state is significantly cleaner when considering our fail states - at any time the cluster manager can attempt an eviction; the cluster manager also can try to repair stuck nodes with the `NodeMonitorService`; and the individual node can detect fail states and attempt internal recovery as well.

File: kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java:45]: Let's leave this in for this PR, as #124 is already good to go once this is merged.

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java:107]: This was reworked per feedback here - https://github.com/slackhq/kaldb/pull/122#discussion_r715819975. 

The `CompletableFuture` does not take a duration as a timeout so this approach is required.

File: kaldb/src/main/java/com/slack/kaldb/chunk/ChunkRollOverException.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/ChunkRollOverException.java:1]: Split into separate PR #126 which is ready for review

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java:32]: Done.

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/caching/CachingChunkManager.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/manager/caching/CachingChunkManager.java]: Done.

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/indexing/IndexingChunkManager.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/manager/indexing/IndexingChunkManager.java]: Done.

File: kaldb/src/main/java/com/slack/kaldb/chunk/manager/IndexingChunkManager.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/chunk/manager/IndexingChunkManager.java:68]: Ah, good call. I had originally fixed this and then aborted due to the PR getting a bit out of hand.

File: kaldb/src/main/java/com/slack/kaldb/metadata/core/KaldbMetadataStore.java
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/metadata/core/KaldbMetadataStore.java]: Done.

**Pull request merged**
